### PR TITLE
Feature: Interactive Mode

### PR DESCRIPTION
**Pull request by agwosdz**
Added command line argument --interactive to use interactive mode for parameter selection to make it more user friendly for novice users :)

image

It also allows for Smashing Enter to keep everything default

Also, made the stars printed in print_version dynamic :)

### CODE REVIEW

**Review by rakuri255 — CHANGES_REQUESTED** [review]

**Comment by rakuri255** [comment]
I like this interavtive mode!

File: src/Settings.py [hunk 1]
```diff
@@ -21,7 +21,8 @@ class Settings:
     ignore_audio = False
     input_file_is_ultrastar_txt = False # todo: to process_data
     keep_cache = False
-
+    interactive_mode = False
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/Settings.py:24]: removed

File: src/UltraSinger.py [hunk 2]
```diff
@@ -699,10 +710,152 @@ def arg_options():
         "keep_cache=",
         "musescore_path=",
         "cookiefile=",
-        "keep_numbers="
+        "keep_numbers=",
+        "interactive"
     ]
     return long, short
 
+def init_settings_interactive() -> Settings:
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py]: Moved to module/init_interactive_mode.py

File: src/UltraSinger.py [hunk 3]
```diff
@@ -592,6 +600,7 @@ def init_settings(argv: list[str]) -> Settings:
     if len(opts) == 0:
         print_help()
         sys.exit()
+    settings.interactive_mode = False
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py]: removed

File: src/UltraSinger.py [hunk 4]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+        "interactive"
     ]
     return long, short
 
+def init_settings_interactive() -> Settings:  
+    ULTRASINGER_HEAD = "[bold green][UltraSinger][/bold green]"  
+    console = Console()
+    console.print(f"{ULTRASINGER_HEAD} [gold]UltraSinger Interactive Mode[/gold]\n")  
+
+    # Input file
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py:718]: I saw that, moved to own module, and split up into separate functions

File: src/UltraSinger.py [hunk 5]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+    # Input file  
+    while True:  
+        input_file = console.input(f"{ULTRASINGER_HEAD} Enter the path to the input file ([green]audio file[/green], [green]Ultrastar txt[/green], or [green]YouTube URL[/green]): ").strip()  
+        if input_file:  
+            settings.input_file_path = input_file  
+            break  
+        else:  
+            console.print(f"{ULTRASINGER_HEAD} [bold red]Error:[/bold red] Input file cannot be empty. Please try again.\n")  
+
+    # Output folder
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py:727]: moved to get_output_folder in init_interactive_mode.py

File: src/UltraSinger.py [hunk 6]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+            # Default model  
+            settings.whisper_model = WhisperModel.LARGE_V2  
+            break  
+        elif model_choice.isdigit() and 1 <= int(model_choice) <= len(whisper_models):  
+            settings.whisper_model = WhisperModel(whisper_models[int(model_choice) - 1])  
+            break  
+        else:  
+            console.print(f"{ULTRASINGER_HEAD} [bold red]Error:[/bold red] Invalid choice. Please select a valid number.\n")  
+
+    # Whisper align model
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py:771]: combined to one function select_model in init_interactive_mode.py

File: src/UltraSinger.py [hunk 7]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+            dirname = os.getcwd()  
+        else:  
+            dirname = os.path.dirname(settings.input_file_path)  
+        settings.output_folder_path = os.path.join(dirname, "output")  
+
+    # Whisper model  
+    whisper_models = [model.value for model in WhisperModel]  
+    console.print(f"\n{ULTRASINGER_HEAD} [bold underline]Available Whisper Models:[/bold underline]\n")  
+
+    # Display options in 4 columns using Table
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py]: With the rewrite, I don't think that is needed anymore

File: src/UltraSinger.py [hunk 8]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+    if output_folder:  
+        settings.output_folder_path = output_folder  
+    else:  
+        if settings.input_file_path.startswith("https:"):  
+            dirname = os.getcwd()  
+        else:  
+            dirname = os.path.dirname(settings.input_file_path)  
+        settings.output_folder_path = os.path.join(dirname, "output")  
+
+    # Whisper model
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py]: combined into select_model in init_interactive_mode.py

File: src/UltraSinger.py [hunk 9]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+            # Default model  
+            settings.demucs_model = DemucsModel.HTDEMUCS  
+            break  
+        elif demucs_choice.isdigit() and 1 <= int(demucs_choice) <= len(demucs_models):  
+            settings.demucs_model = DemucsModel(demucs_models[int(demucs_choice) - 1])  
+            break  
+        else:  
+            console.print(f"{ULTRASINGER_HEAD} [bold red]Error:[/bold red] Invalid choice. Please select a valid number.\n")  
+
+    # Additional options
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py]: Agreed, all additional options are handled by one function.
configure_additional_options in init_interactive_mode.py

File: src/UltraSinger.py [hunk 10]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+    # Whisper align model  
+    settings.whisper_align_model = console.input(f"\n{ULTRASINGER_HEAD} Enter the [green]Whisper Align Model[/green] or leave empty for none: ").strip()  
+    if not settings.whisper_align_model:  
+        settings.whisper_align_model = None  
+
+    # DEMUCS model  
+    demucs_models = [model.value for model in DemucsModel]  
+    console.print(f"\n{ULTRASINGER_HEAD} [bold underline]Available Demucs Models:[/bold underline]\n")  
+
+    # Display options in 4 columns using Table
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py]: resolved with rewrite

File: src/UltraSinger.py [hunk 11]
```diff
@@ -699,10 +710,152 @@ def arg_options():
...
+            break  
+        else:  
+            console.print(f"{ULTRASINGER_HEAD} [bold red]Error:[/bold red] Invalid choice. Please select a valid number.\n")  
+
+    # Whisper align model  
+    settings.whisper_align_model = console.input(f"\n{ULTRASINGER_HEAD} Enter the [green]Whisper Align Model[/green] or leave empty for none: ").strip()  
+    if not settings.whisper_align_model:  
+        settings.whisper_align_model = None  
+
+    # DEMUCS model
```
[the remark this answers is not in the document]
Comment (agwosdz) [reply src/UltraSinger.py]: solved in combined select_model function in init_interactive_mode.py

**Review by rakuri255 — APPROVED** [review]

**Comment by rakuri255** [comment]
@agwosdz can you upmerge it? Than i reintegrste it 😀

**Pull request merged**